src/utils/batchDiscordRequests.ts

```diff
@@ -0,0 +1,133 @@
+import JSONResponse from "./JsonResponse";
+import { addDelay, convertSecondsToMillis } from "./timeUtils";
+export const DISCORD_HEADERS = {
+  RATE_LIMIT_RESET_AFTER: "X-RateLimit-Reset-After",
+  RATE_LIMIT_REMAINING: "X-RateLimit-Remaining",
+  RETRY_AFTER: "Retry-After",
+};
+
+const MAX_RETRY = 1;
+const LIMIT_BUFFER = 0.2;
+
+interface RequestDetails {
+  retries: number;
+  request: () => Promise<Response>;
+  index: number;
+}
+interface ResponseDetails {
+  response: Response;
+  data: RequestDetails;
+}
+
+const parseRateLimitRemaining = (response: Response) => {
+  let rateLimitRemaining = Number.parseInt(
+    response.headers.get(DISCORD_HEADERS.RATE_LIMIT_REMAINING) || "0"
+  );
+  rateLimitRemaining = Math.floor(rateLimitRemaining * (1 - LIMIT_BUFFER));
+  return rateLimitRemaining;
+};
+
+const parseResetAfter = (response: Response) => {
+  let resetAfter = Number.parseFloat(
+    response.headers.get(DISCORD_HEADERS.RATE_LIMIT_RESET_AFTER) || "0"
+  );
+  resetAfter = Math.ceil(resetAfter);
+  return resetAfter;
+};
+
+export const batchDiscordRequests = async (
+  requests: { (): Promise<Response> }[]
+): Promise<Response[]> => {
+  try {
+    const requestsQueue: RequestDetails[] = requests.map((request, index) => {
+      return {
+        retries: 0,
+        request: request,
+        index: index,
+      };
+    });
+
+    const responseList: Response[] = new Array(requestsQueue.length);
+    let resetAfter = 0;
+    let nextMinimumResetAfter = Infinity;
+    let rateLimitRemaining = 1;
+    let nextMinimumRateLimitRemaining = Infinity;
+
+    const handleResponse = async (
+      response: JSONResponse,
+      data: RequestDetails
+    ): Promise<void> => {
+      if (response.ok) {
+        nextMinimumResetAfter = Math.min(
+          nextMinimumResetAfter,
+          parseResetAfter(response)
+        );
+        nextMinimumRateLimitRemaining = Math.min(
+          nextMinimumRateLimitRemaining,
+          parseRateLimitRemaining(response)
+        );
+
+        responseList[data.index] = response;
+      } else {
+        nextMinimumResetAfter = Math.min(
+          nextMinimumResetAfter,
+          parseResetAfter(response)
+        );
+        rateLimitRemaining = 0;
+        if (data.retries >= MAX_RETRY) {
+          responseList[data.index] = response;
+        } else {
+          data.retries++;
+          requestsQueue.push(data);
+        }
+      }
+    };
+
+    const executeRequest = async (
```

Comment:
When an error occurs during the request execution, the code logs a message with the error information but doesn't retain the original request details within the error object. This makes it difficult to identify the specific request that caused the error and analyze the root cause.

Reply:
- Here we are ignoring and only logging the first error that occurs and retrying the same request. 
- If the request fails for the second time we are returning the status as failure for the given input through the api response.

Please let me know if there are any other suggestions. 

<img width="774" alt="Screenshot 2023-12-10 at 5 26 31 AM" src="https://github.com/Real-Dev-Squad/discord-slash-commands/assets/98796547/21e729bd-fd79-4ea5-9853-74aa2af973f1">